### PR TITLE
Dotnet template ci

### PR DESCRIPTION
**Pull request by RobotOptimist**
#14 
this creates a hello world project and a template pack project

## ci changes

I've changed the actions/checkout from @v1 to @v2 to get the benefit of the improvements from that github action
I'm also fetching all commit history so nbgv works properly

there are several new steps, I'm pulling in the NerdBank github action to get easy access to all of the build variables, this also abstracts away the installation of the nbgv tool

I've added 2 additional steps that will only run on push to master

1. I'm updating the hello world project featherhttp version using a nerdbank variable and --no restore
2. I'm packing the template pack project.

## file changes

I've added 

* a basic helloworld program.cs
* a nuget.config with a reference to the davidfowl feedz.io
* a csproj file with the above files included as content
* a template.json file to describe the template
* a template csproj file that packages the above files into a template project
* a version.json file for the template csproj to maintain an independent version. [Why?](https://gist.github.com/RobotOptimist/ac8061dcdf2085bbbbecd551684219e5)
* also modified .gitignore to ignore dist/ and .nuxt/ folders

## usage
(I'm writing this to be copy pastable into a readme)

This feedz.io: `https://f.feedz.io/davidfowl/featherhttp/nuget/index.json` needs to be added as a global nuget source that can be done via the [nuget cli](https://docs.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-sources) or [visual studio](https://docs.microsoft.com/en-us/nuget/consume-packages/install-use-packages-visual-studio#package-sources).

Once the nuget source is added added you can install the template like so `dotnet new -i FeatherHttp.Templates`
and then create your new project like this `dotnet new featherhttp -n MyProject`
Then you can `cd MyProject` and view/run the project assets.

Thanks!

### CODE REVIEW

File: template/Content/FeatherHttpContent.csproj [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <RootNamespace>featherhttp_template</RootNamespace>
```
Comment (davidfowl) [inline template/Content/FeatherHttpContent.csproj]: Why is this part of the template csproj?

File: template/Content/FeatherHttpContent.csproj [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="FeatherHttp" Version="0.1.34-alpha.g005ace427a" />
+    <Content Include="Program.cs" />
```
Comment (davidfowl) [inline template/Content/FeatherHttpContent.csproj]: Why are these in the csproj?

File: template/Content/FeatherHttpContent.csproj (hunk 1 above)
Comment (RobotOptimist) [reply template/Content/FeatherHttpContent.csproj]: RootNamespace is not required and is also incorrect. I've removed the RootNamespace in my most recent commit 'cleaning up template csproj'

Please let me know if you would like additional changes here

File: template/Content/FeatherHttpContent.csproj (hunk 2 above)
Comment (RobotOptimist) [reply template/Content/FeatherHttpContent.csproj]: Program.cs and nuget.config are not required in the ItemGroup - although the csproj vscode extension keeps prompting me to add them XD. I've now removed them in my most recent commit 'cleaning up template csproj'

Please let me know if you would like additional changes here

File: template/Project/FeatherHttpProject.csproj [hunk 3]
```diff
@@ -9,7 +9,7 @@
     <TargetFramework>netcoreapp3.1</TargetFramework>    
     <IncludeContentInPack>true</IncludeContentInPack>
     <IncludeBuildOutput>false</IncludeBuildOutput>
-    <ContentTargetFolders></ContentTargetFolders>    
+    <ContentTargetFolders>content/content</ContentTargetFolders>
```
Comment (RobotOptimist) [inline template/Project/FeatherHttpProject.csproj]: I've added content/content here to try to address an issue where when applying the installed template it goes into a root folder of `netcoreapp31` 
My testing indicated that adding this line here changed `netcoreapp31` to `content` which is perhaps marginally better but not ideal. I've tried multiple experiments and tests to try to make this behave the way I think it should, i.e. if you `dotnet new featherhttp -o .` it should add the project files into the directory instead of in a nested directory. 
Currently, this is not behaving the way I expect it to but I've been unable to resolve it. 
I've tried:

* modifying the output directory on build
* omitting any value in ContentTargetFolders (results in `netcoreapp31` as the nested directory)
* using powershell / build tasks to organize project assets similar to the tutorial implementation.

and a few more variations on the above
I've consulted the following resources to seek guidance on fixing this problem:

* https://docs.microsoft.com/en-us/dotnet/core/tutorials/cli-templates-create-template-pack
* https://docs.microsoft.com/en-us/nuget/reference/msbuild-targets
* https://github.com/dotnet/dotnet-template-samples
* https://devblogs.microsoft.com/dotnet/how-to-create-your-own-templates-for-dotnet-new/
* https://github.com/petabridge/petabridge-dotnet-new/tree/dev/src/templates //lots of ideas from this on how to add options for this template later
* https://github.com/dotnet/templating/wiki/Available-templates-for-dotnet-new //looked at several of these but not all of them - could not find guidance
* https://docs.microsoft.com/en-us/dotnet/core/tools/csproj

Do you know where I can look further to solve this problem?

File: template/Content/FeatherHttpContent.csproj (hunk 2 above)
Comment (davidfowl) [reply template/Content/FeatherHttpContent.csproj]: I don't see the changes yet

Comment (RobotOptimist) [reply template/Content/FeatherHttpContent.csproj]: Apologies, it seems I removed the RootNamespace but not these files as content items. I've now done so in the most recent commit

**Comment by davidfowl** [comment]
So things mostly look good, except for the hard coded version in source control. Where did we get with that?

**Comment by RobotOptimist** [comment]
@davidfowl right on, your talking about the PackageReference with the hardcoded version of 0.1.34-alpha.g005ace427a
I can replace that with 0.0.0 or 0.0.0-ci.changes as the version is updated in the ci.
I had experimented with updating the version in the ci and then committing & pushing the changes (also in the ci) back to master - this can be made to work but it's not great. It tends to cause issues with the git versioning. 
My thought is that if it's 0.0.0 then even if it's not perfect, it's at least not misleading since probably no one will think that's the actual version. Let me know if that works

**Comment by davidfowl** [comment]
@RobotOptimist I was thinking more about having an variable checked in, like `$(TemplateVersion)` that would be specified at build time when making the template.

**Comment by RobotOptimist** [comment]
@davidfowl that variable $(PackageVersion) is [created](https://github.com/AArnott/Nerdbank.GitVersioning/blob/a7d04e152a42c3708a172721c59178a97930a485/src/Nerdbank.GitVersioning.Tasks/build/Nerdbank.GitVersioning.targets#L100) by NerdBank but it does not update the package version in a csproj embedded as content in a template pack csproj. 
What we get instead is this after installing and using the template:

```
  <ItemGroup>
    <PackageReference Include="FeatherHttp" Version="$(PackageVersion)" />
  </ItemGroup>
```
However, we can still achieve this without using msbuild by using powershell string replace.
Something like this:
```
((Get-Content -path ./template/TemplateContent/FeatherHttpProject.csproj -Raw) -replace "TheNugetPackageVersion", $env:NBGV_NuGetPackageVersion) | Set-Content -Path ./template/TemplateContent/FeatherHttpProject.csproj
```
I mean, I can read it in as XML and do the update that way as well, but this works pretty well tbh

**Comment by davidfowl** [comment]
The only thing I wanted to do is avoid that hardcoded version being checked in.

**Comment by RobotOptimist** [comment]
@davidfowl No problem! I've removed that.

File: template/Content/FeatherHttpContent.csproj [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="FeatherHttp" Version="TheNugetPackageVersion" />
```
Comment (davidfowl) [inline template/Content/FeatherHttpContent.csproj]: Maybe a variable? Even if its not specified to make the project file valid syntax?

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -4,20 +4,39 @@ on: [push, pull_request]
...
   build:
-
     runs-on: windows-latest
 
     steps:
-    - uses: actions/checkout@v1
-      
+    - uses: actions/checkout@v2
+    - run: |
+        git fetch --prune --unshallow origin HEAD
```
Comment (davidfowl) [inline .github/workflows/ci.yaml]: ? Why this change?

File: template/Content/FeatherHttpContent.csproj (hunk 4 above)
Comment (RobotOptimist) [reply template/Content/FeatherHttpContent.csproj]: I've changed this to use the $(PackageVersion) supplied by the nerdbrain - even though msbuild is not replacing this on a `dotnet pack` it should still work if you do a straight build of that project

File: template/Content/Program.cs [hunk 6]
```diff
@@ -0,0 +1,25 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+
+namespace FeatherHttp.Templates.HelloWorld
```
Comment (davidfowl) [inline template/Content/Program.cs]: This changes to the user's namespace right?

File: .gitignore [hunk 7]
```diff
@@ -13,6 +13,8 @@ BenchmarkDotNet.Artifacts/
 .gradle/
 src/SignalR/clients/**/dist/
 modules/
+dist/
```
Comment (davidfowl) [inline .gitignore]: what are these?

File: .github/workflows/ci.yaml (hunk 5 above)
Comment (RobotOptimist) [reply .github/workflows/ci.yaml]: checkout@v2 seems to have some improvements - but reading through them it's not something the project needs at the moment - apologies for exceeding the scope on this piece. 
I've removed this change

File: template/Content/Program.cs (hunk 6 above)
Comment (RobotOptimist) [reply template/Content/Program.cs]: correct when the user provides the -n this will change also
`dotnet new featherhttp -n My.Awesome.Project` will result in a new directory of that name and the namespace will change as well

File: template/Project/version.json [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (davidfowl) [inline template/Project/version.json]: Is this part of the generated template? Why is it here?

File: .gitignore (hunk 7 above)
Comment (RobotOptimist) [reply .gitignore]: developing in the same repo as docs where these build output folders are ignored, this was expedient in order to work on both tasks

File: template/Project/version.json (hunk 8 above)
Comment (RobotOptimist) [reply template/Project/version.json]: Hi David, I anticipated this question in my initial message on this pull request. 

The dotnet new templating engine does not allow prerelease templates to be installed. 
Here's the issue: https://github.com/dotnet/templating/issues/1423

Therefore the template needs to be a stable release with no -alpha or similar tags.

As there is already a version.json in the project which stipulates the featherhttp project is in alpha it was necessary to add a new version.json file just for the template project in order to overcome this issue. 

Once FeatherHttp is out of prerelease and you want to have the template version and the FeatherHttp version match then this can be removed/reconsidered

Comment (davidfowl) [reply template/Project/version.json]: https://devblogs.microsoft.com/aspnet/blazor-webassembly-3-2-0-preview-1-release-now-available/

```
dotnet new -i Microsoft.AspNetCore.Blazor.Templates::3.2.0-preview1.20073.1
```
What am I missing?

Comment (RobotOptimist) [reply template/Project/version.json]: Nothing at all, I'm incorrect on this. I didn't know you could request the version like that upon installing a new dotnet template - I'll remove the version.json and users will need to stipulate the version as above in order for it to work

File: template/Content/.template.config/template.json [hunk 9]
```diff
@@ -0,0 +1,16 @@
+{
+    "$schema": "http://json.schemastore.org/template",
+    "author": "David Fowler",
+    "classifications": [ "Web", "ASP.NET", "FeatherHttp" ],
+    "identity": "FeatherHttp.Templates",
+    "name": "FeatherHttp Templates",
+    "description": "Am ultra lightweight ASP.NET project starter",
```
Comment (davidfowl) [inline template/Content/.template.config/template.json]: typo: Am -> An

Comment (davidfowl) [reply template/Content/.template.config/template.json]: Remove ASP.NET, put .NET Core

Comment (davidfowl) [reply template/Content/.template.config/template.json]: Lightweight low ceremony API for ASP.NET Core services.

Comment (RobotOptimist) [reply template/Content/.template.config/template.json]: Okay, I've fixed that typo

**Comment by davidfowl** [comment]
Thanks for putting up with my nits 😄
